File: zw_brain/domain/data_quality.py
"""数据质量分类器 — 真实导入单「用途/使用事由」脏值的机械化判定（单一规则源）。

客户试用反馈（0604，业务方试用反馈 + 截图实证）：真实导入申请单的「用途」列出现「测试」「167」
「169,167」这类脏值裸奔在需方/办理视图。乔布斯裁决（本组）：

  1) 列表渲染时脏值**降级显示**（「未填写用途」次要样式），不让噪音污染需方视图；
  2) 「用途缺失/无效」的单子计入**供方数据质量队列**（数据质量是供方的待办，不是需方的噪音）。

本模块是「什么算脏值」的**唯一规则源**（Python 权威 + 前端 dataQuality.ts 镜像 +
tests/test_data_quality_classifier.py 守一致）。规则机械化，不靠散文：

  - 空 / 纯空白              → empty
  - 纯数字 / 数字+分隔符      → numeric_only（如 "167" "169,167" "12-3"）
  - 占位测试词（枚举）        → placeholder（如 "测试" "test" "demo" "无" "暂无" "1"）
  - 过短（去空白后 < 2 字符）  → too_short

任一非空类目 → 视为脏值（``is_dirty`` True），列表降级 + 进供方质量队列。
"""

from __future__ import annotations

import re

__all__ = [
    "classify_purpose",
    "is_dirty_purpose",
    "purpose_from_payload",
    "MISSING_PURPOSE_LABEL",
]

# 列表降级文案（政务白话，天然过 R12：无工程术语 / 无 snake_case / 无 HTTP 码）。
MISSING_PURPOSE_LABEL = "未填写用途"

# 占位/测试词枚举（去空白 + 转小写后精确匹配）。保持克制——只收真见过的占位串，
# 避免误伤合法短用途。新增脏值类型优先扩这里或下面的规则，不在调用方散落 if。
_PLACEHOLDER_TOKENS: frozenset[str] = frozenset(
    {
        "测试",
        "test",
        "测试数据",
        "demo",
        "演示",
        "无",
        "暂无",
        "待填",
        "待补充",
        "n/a",
        "na",
        "null",
        "none",
        "-",
        "—",
        "1",
        "0",
    }
)

# 纯数字 / 数字 + 常见分隔符（逗号 / 顿号 / 横线 / 斜杠 / 点 / 空格）。
_NUMERIC_ONLY = re.compile(r"^[\d\s,，、\-/\.]+$")


def classify_purpose(raw: str | None) -> str:
    """返回脏值类目：clean | empty | numeric_only | placeholder | too_short。

    clean = 合法用途（不降级）；其余 = 脏值（降级 + 进供方质量队列）。
    """
    text = (raw or "").strip()
    if not text:
        return "empty"
    lowered = text.lower()
    if lowered in _PLACEHOLDER_TOKENS:
        return "placeholder"
    if _NUMERIC_ONLY.match(text):
        return "numeric_only"
    # 去掉所有空白后长度 < 2（如单字符残值），视为信息量不足。
    if len(re.sub(r"\s+", "", text)) < 2:
        return "too_short"
    return "clean"


def is_dirty_purpose(raw: str | None) -> bool:
    """脏值判定：任一非 clean 类目 → True（列表降级 + 计入供方质量队列）。"""
    return classify_purpose(raw) != "clean"


def purpose_from_payload(payload: dict | None) -> str:
    """application_record.payload_json → 原始用途文本（多 legacy 字段名兜底，口径单源）。

    与 discovery_snapshot_projection / workbench_backlog_projection 共用，避免两处
    各写一遍字段兜底序导致脏值计数口径漂移。
    """
    p = payload or {}
    return str(
        p.get("purpose") or p.get("use_reason") or p.get("apply_basis") or p.get("use_item") or ""
    )

File: zw-brain-web/src/lib/dataQuality.ts
/**
 * 数据质量分类器（前端镜像）— 真实导入单「用途」脏值的机械化判定。
 *
 * 权威规则源在后端 `zw_brain/domain/data_quality.py`；本文件是前端镜像，
 * 用于列表渲染时**降级脏值**（不裸奔）。两侧由 tests/test_data_quality_classifier.py
 * 的固定样例（测试 / 167 / 169,167 / 空）守住一致——改一侧规则须同改另一侧。
 *
 * 客户试用反馈（0604，业务方试用反馈 + 截图）：「用途列出现『测试』『167』这类脏值」。
 * 裁决：列表脏值降级显示「未填写用途」次要样式，不再把历史脏值伪装成供数据页待办。
 */

/** 列表降级文案（与后端 MISSING_PURPOSE_LABEL 一致）。 */
export const MISSING_PURPOSE_LABEL = '未填写用途';

export type PurposeQuality = 'clean' | 'empty' | 'numeric_only' | 'placeholder' | 'too_short';

// 占位/测试词枚举（去空白 + 转小写后精确匹配）。与后端 _PLACEHOLDER_TOKENS 一致。
const PLACEHOLDER_TOKENS: ReadonlySet<string> = new Set([
  '测试',
  'test',
  '测试数据',
  'demo',
  '演示',
  '无',
  '暂无',
  '待填',
  '待补充',
  'n/a',
  'na',
  'null',
  'none',
  '-',
  '—',
  '1',
  '0',
]);

// 纯数字 / 数字 + 常见分隔符（逗号 / 顿号 / 横线 / 斜杠 / 点 / 空格）。
const NUMERIC_ONLY = /^[\d\s,，、\-/.]+$/;

/** 返回脏值类目；clean = 合法用途（不降级），其余 = 脏值。 */
export function classifyPurpose(raw: string | null | undefined): PurposeQuality {
  const text = (raw ?? '').trim();
  if (!text) return 'empty';
  const lowered = text.toLowerCase();
  if (PLACEHOLDER_TOKENS.has(lowered)) return 'placeholder';
  if (NUMERIC_ONLY.test(text)) return 'numeric_only';
  if (text.replace(/\s+/g, '').length < 2) return 'too_short';
  return 'clean';
}

/** 脏值判定：非 clean → true（列表降级）。 */
export function isDirtyPurpose(raw: string | null | undefined): boolean {
  return classifyPurpose(raw) !== 'clean';
}

/**
 * 列表展示用途：脏值 → 降级文案（次要样式由调用方据 dirty 标志渲染）。
 * 返回 { text, dirty }：text 是该显示的字符串，dirty 决定是否打次要灰样式。
 */
export function displayPurpose(raw: string | null | undefined): { text: string; dirty: boolean } {
  if (isDirtyPurpose(raw)) return { text: MISSING_PURPOSE_LABEL, dirty: true };
  return { text: (raw ?? '').trim(), dirty: false };
}
